client/asset/dcr/dcr.go

```diff
@@ -4521,3 +4540,190 @@ func float64PtrStr(v *float64) string {
 	}
 	return strconv.FormatFloat(*v, 'f', 8, 64)
 }
+
+func (dcr *ExchangeWallet) ConfirmRedemption(coinID dex.Bytes, redemption *asset.Redemption, feeSuggestion uint64) (*asset.ConfirmRedemptionStatus, error) {
+	txHash, _, err := decodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+
+	var secretHash [32]byte
+	copy(secretHash[:], redemption.Spends.SecretHash)
+	dcr.mempoolRedeemsMtx.RLock()
+	mRedeem, have := dcr.mempoolRedeems[secretHash]
+	dcr.mempoolRedeemsMtx.RUnlock()
+
+	var deleteMempoolRedeem bool
+	defer func() {
+		if deleteMempoolRedeem {
+			dcr.mempoolRedeemsMtx.Lock()
+			delete(dcr.mempoolRedeems, secretHash)
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+	}()
+
+	var abandonTx bool
+	tx, err := dcr.wallet.GetTransaction(dcr.ctx, txHash)
+	if err == nil {
+		if have && mRedeem.txHash == *txHash {
+			if tx.Confirmations == 0 && time.Now().After(mRedeem.firstSeen.Add(maxRedeemMempoolAge)) {
+				// Transaction has been sitting in the mempool
+				// for a long time now. Attempt to abandon.
+				abandonTx = true
+				if err := dcr.wallet.AbandonTransaction(dcr.ctx, txHash); err != nil {
+					dcr.log.Warnf("Unable to abandon redemption mempool transaction %v: %v.", txHash, err)
+				} else {
+					dcr.log.Infof("Abandoned redemption mempool transaction: %v.", txHash)
+					deleteMempoolRedeem = true
+				}
+			}
+		} else {
+			// Will hit this if node was restarted with an actively
+			// redeeming swap.
+			dcr.mempoolRedeemsMtx.Lock()
+			dcr.mempoolRedeems[secretHash] = &mempoolRedeem{txHash: *txHash, firstSeen: time.Now()}
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+		if !abandonTx {
+			if tx.Confirmations >= requiredRedeemConfirms {
+				deleteMempoolRedeem = true
+			}
+			return &asset.ConfirmRedemptionStatus{
+				Confs:  uint64(tx.Confirmations),
+				Req:    requiredRedeemConfirms,
+				CoinID: coinID,
+			}, nil
+		}
+	}
+
+	if !abandonTx && !errors.Is(err, asset.CoinNotFoundError) {
+		return nil, fmt.Errorf("problem searching for redemption transaction %s: %w", txHash, err)
+	}
+
+	// Redemption transaction is missing from the point of view of our node!
+	// Unlikely, but possible it was redeemed by another transaction. We
+	// assume a contract past its locktime cannot make it here, so it must
+	// not be refunded. Check if the contract is still an unspent output.
+
+	swapHash, vout, err := decodeCoinID(redemption.Spends.Coin.ID())
+	if err != nil {
+		return nil, err
+	}
+
+	var spent bool
+	_, err = dcr.wallet.UnspentOutput(dcr.ctx, swapHash, vout, 0)
+	if err != nil {
+		if errors.Is(err, asset.CoinNotFoundError) {
+			spent = true
```

Comment:
The naming of `spent` here is misleading here since in SPV mode we can't know it's spent based on `UnspentOutput`.  It's more like `maybeSpent`.  In SPV mode, it's _expected_ to not know if the contract is spent until you do a cfilters scan.  See the weak guarantees documented for `UnspentOutput` in the `asset.Wallet` interface docs.

Also, you do follow this with a block filters scan to determine, but why don't we use `lookupTxOutput` instead of using `UnspentOutput` directly before the scan like in the other places we do this check like in `refundTx`?  It's a bit smarter when not in spv mode.

Reply:
Using `lookupTxOutput` now.